src/System.Private.CoreLib/shared/System/Collections/Generic/List.cs

```diff
@@ -339,11 +341,19 @@ public List<TOutput> ConvertAll<TOutput>(Converter<T, TOutput> converter)
             }
 
             List<TOutput> list = new List<TOutput>(_size);
-            for (int i = 0; i < _size; i++)
+            Debug.Assert((uint)_size <= (uint)list._items.Length);
+            list._size = _size;
+            // Use Span to avoid covarience checks from using List.
+            // We can use the ref overload as the list is a local and not subject to data races.
+            Span<TOutput> output = new Span<TOutput>(ref Unsafe.As<byte, TOutput>(ref list._items.GetRawSzArrayData()), _size);
+
+            int i = 0;
+            // Use Span to avoid additional bounds checks from using List
```

Comment:
Nit: here and in similar cases, could you make the comment just be at the end of the foreach line? e.g.
```C#
foreach (T item in new ReadOnlySpan<T>(_items, 0, _size)) // using span to avoid unnecessary bounds checks
```
To me personally, that helps to highlight that the comment is explaining a choice made on that line rather than explaining something about the whole loop.